ffi/src/winscard/system_scard/card.rs

```diff
@@ -0,0 +1,403 @@
+use std::borrow::Cow;
+use std::mem::size_of;
+use std::ptr::null_mut;
+use std::slice::from_raw_parts;
+
+#[cfg(target_os = "windows")]
+use ffi_types::winscard::functions::SCardApiFunctionTable;
+use ffi_types::winscard::ScardIoRequest;
+#[cfg(target_os = "windows")]
+use ffi_types::winscard::{ScardContext, ScardHandle};
+use num_traits::ToPrimitive;
+use winscard::winscard::{
+    AttributeId, ControlCode, IoRequest, Protocol, ReaderAction, ShareMode, Status, TransmitOutData, WinScard,
+};
+use winscard::{Error, ErrorKind, WinScardResult};
+
+use super::parse_multi_string_owned;
+use crate::winscard::buf_alloc::SCARD_AUTOALLOCATE;
+#[cfg(not(target_os = "windows"))]
+use crate::winscard::pcsc_lite::functions::PcscLiteApiFunctionTable;
+#[cfg(not(target_os = "windows"))]
+use crate::winscard::pcsc_lite::{initialize_pcsc_lite_api, ScardContext, ScardHandle};
+
+/// Represents a state of the current `SystemScard`.
+#[derive(Copy, Clone)]
+enum HandleState {
+    /// The card is not connected or has been disconnected.
+    Disconnected,
+    /// The card is connected and ready to use.
+    Connected(ScardHandle),
+}
+
+pub struct SystemScard {
+    h_card: HandleState,
+    h_card_context: ScardContext,
+    #[cfg(target_os = "windows")]
+    api: SCardApiFunctionTable,
+    #[cfg(not(target_os = "windows"))]
+    api: PcscLiteApiFunctionTable,
+}
+
+impl SystemScard {
+    pub fn new(h_card: ScardHandle, h_card_context: ScardContext) -> WinScardResult<Self> {
+        if h_card == 0 {
+            return Err(Error::new(
+                ErrorKind::InvalidParameter,
+                "Scard handle can not be a zero.",
+            ));
+        }
+
+        if h_card_context == 0 {
+            return Err(Error::new(
+                ErrorKind::InvalidParameter,
+                "Scard context handle can not be a zero.",
+            ));
+        }
+
+        Ok(Self {
+            h_card: HandleState::Connected(h_card),
+            h_card_context,
+            #[cfg(target_os = "windows")]
+            api: super::init_scard_api_table()?,
+            #[cfg(not(target_os = "windows"))]
+            api: initialize_pcsc_lite_api()?,
+        })
+    }
+
+    fn h_card(&self) -> WinScardResult<ScardHandle> {
+        if let HandleState::Connected(handle) = self.h_card {
+            Ok(handle)
+        } else {
+            Err(Error::new(
+                ErrorKind::InvalidHandle,
+                "Smart card is not connected or has been disconnected.",
```

Comment:
style: https://github.com/Devolutions/IronRDP/blob/63743e9fcc5e3adf37e4745cc1e3ba7858e16c39/STYLE.md#formatting-of-error-messages

```suggestion
                "smart card is not connected or has been disconnected",
```

Reply:
Done.
I forgot about this again :persevere: